src/main/java/com/google/devtools/build/lib/runtime/BlazeOptionHandler.java

```diff
@@ -194,12 +230,15 @@ void parseRcOptions(
                   : String.format(
                       "Reading rc options for '%s' from %s",
                       commandAnnotation.name(), rcArgs.getRcFile());
-          rcfileNotes.add(
-              String.format(
-                  "%s:\n  %s'%s' options: %s",
-                  source, inherited, commandToParse, Joiner.on(' ').join(rcArgs.getArgs())));
+          String rcFileNote = String.format(
+              "%s:\n  %s'%s' options: %s",
+              source, inherited, commandToParse, Joiner.on(' ').join(rcArgs.getArgs()));
+          if (!ignoredArgs.isEmpty()) {
+            rcFileNote += String.format("\n  Ignored as unsupported by '%s': %s",
+                commandAnnotation.name(), Joiner.on(' ').join(ignoredArgs));
+          }
+          rcfileNotes.add(rcFileNote.toString());
```

Comment:
Minor issue here detected by an internal test when I tried to import. Adding the note after the call to `optionsParser.parse()` doesn't work because that call may throw `OptionsParsingException`, and we won't get a note for the illegal option. Seems like that test case should be open sourced. Anyway, I'm changing this to:


```java
for (RcChunkOfArgs rcArgs : commandToRcArgs.get(commandToParse)) {
  if (!rcArgs.getArgs().isEmpty()) {
    String inherited = commandToParse.equals(commandAnnotation.name()) ? "" : "Inherited ";
    String source =
        rcArgs.getRcFile().equals("client")
            ? "Options provided by the client"
            : String.format(
                "Reading rc options for '%s' from %s",
                commandAnnotation.name(), rcArgs.getRcFile());
    rcfileNotes.add(
        String.format(
            "%s:\n  %s'%s' options: %s",
            source, inherited, commandToParse, Joiner.on(' ').join(rcArgs.getArgs())));
  }
  if (commandToParse.equals(COMMON_PSEUDO_COMMAND)) {
    // Pass in options data for all commands supported by the runtime so that options that
    // apply to some but not the current command can be ignored.
    //
    // Important note: The consistency checks performed by
    // OptionsParser#getFallbackOptionsData ensure that there aren't any two options across
    // all commands that have the same name but parse differently (e.g. because one accepts
    // a value and the other doesn't). This means that the options available on a command
    // limit the options available on other commands even without command inheritance. This
    // restriction is necessary to ensure that the options specified on the "common"
    // pseudo command can be parsed unambiguously.
    ImmutableList<String> ignoredArgs =
        optionsParser.parseWithSourceFunction(
            PriorityCategory.RC_FILE,
            o -> rcArgs.getRcFile(),
            rcArgs.getArgs(),
            OptionsParser.getFallbackOptionsData(allOptionsClasses));
    if (!ignoredArgs.isEmpty()) {
      // Append richer information to the note.
      int index = rcfileNotes.size() - 1;
      String note = rcfileNotes.get(index);
      note +=
          String.format(
              "\n  Ignored as unsupported by '%s': %s",
              commandAnnotation.name(), Joiner.on(' ').join(ignoredArgs));
      rcfileNotes.set(index, note);
    }
  } else {
    optionsParser.parse(PriorityCategory.RC_FILE, rcArgs.getRcFile(), rcArgs.getArgs());
  }
}
```

Reply:
Thanks, the change looks good.

Reply:
With that out of the way, tests are now passing. But apparently we have more internal code outside of blaze that depends on the options parsing libraries, and I need to fix them up too. Please bear with me.